### PR TITLE
[WIP] Symfony3

### PR DESCRIPTION
**Pull request by hacfi**

### CODE REVIEW

**Comment by hacfi** [comment]
@dbu Those are the changes I made to make it work with my current (not online yet) Symfony 3.0 project and no issues for the last 2 weeks. I’m gonna look into the deprecation calls and check if there are more changes required in this repository.

File: Tests/Unit/Routing/DynamicRouterTest.php [hunk 1]
```diff
@@ -89,6 +91,8 @@ public function testMatchRequest()
      */
     public function testMatchNoRequest()
     {
+        $this->markTestSkipped('Test for deprecated match() method');
```
Comment (hacfi) [inline Tests/Unit/Routing/DynamicRouterTest.php]: See https://github.com/symfony-cmf/Routing/blob/master/DynamicRouter.php#L218

Should I just remove those tests?

Comment (WouterJ) [inline Tests/Unit/Routing/DynamicRouterTest.php]: Mark them as legacy and they won't be marked as error. You can do it by adding `@group legacy` to the method PHPdoc.

File: Tests/Resources/app/config/cmf_routing.phpcr.yml [hunk 2]
```diff
@@ -3,4 +3,5 @@ cmf_routing:
         persistence:
             phpcr:
                 enabled: true
-                route_basepath: /test/routing
+                route_basepath:
```
Comment (WouterJ) [inline Tests/Resources/app/config/cmf_routing.phpcr.yml]: You can pass a string once you use the new setting name: `route_basepaths`

File: composer.json [hunk 3]
```diff
@@ -14,7 +14,7 @@
     "require": {
         "php": "^5.3.9|^7.0",
         "symfony-cmf/routing": "~1.2",
-        "symfony/framework-bundle": "~2.3"
+        "symfony/framework-bundle": "~2.3|~3.0"
```
Comment (WouterJ) [inline composer.json]: as we're now using the `placeholder` option, 2.3 has to be 2.6. As 2.6 is no longer supported, let's make it 2.7.

File: Admin/RouteAdmin.php [hunk 4]
```diff
@@ -146,7 +146,7 @@ protected function configureFieldsForDefaults($dynamicDefaults)
             '_controller' => array('_controller', Sf2CompatUtil::getFormTypeName('text'), array('required' => false)),
             '_template' => array('_template', Sf2CompatUtil::getFormTypeName('text'), array('required' => false)),
             'type' => array('type', Sf2CompatUtil::getFormTypeName('cmf_routing_route_type'), array(
-                'empty_value' => '',
+                'placeholder' => '',
```
Comment (hacfi) [inline Admin/RouteAdmin.php]: This causes `PHP: 5.3 - COMPOSER_FLAGS="--prefer-lowest"` and `PHP: 5.6 SYMFONY_VERSION=2.3.*` to fail (see https://github.com/symfony/symfony/blob/2.8/UPGRADE-2.6.md#form).

What is the plan for Symfony 2.3 support? Do I have to add a fix for that (in `Sf2CompatUtil`) or will support be dropped anyway?

**Comment by hacfi** [comment]
Thanks for the feedback @WouterJ. Updated the PR accordingly.

File: Admin/RouteAdmin.php (hunk 4 above)
Comment (hacfi) [inline Admin/RouteAdmin.php]: As @WouterJ suggested I changed the requirements to `~2.7`.

**Comment by hacfi** [comment]
I tried adding Symfony 3.0.* to the test matrix (https://travis-ci.org/symfony-cmf/RoutingBundle/jobs/105928993) but `sonata-project/doctrine-phpcr-admin-bundle 1.1.0 requires symfony/framework-bundle ~2.3`: https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/blob/master/composer.json#L22 to that fails at the moment.

The `COMPOSER_FLAGS="--prefer-lowest"` build (https://travis-ci.org/symfony-cmf/RoutingBundle/jobs/105928991) fails with `Argument 5 passed to Sonata\DoctrinePHPCRAdminBundle\Tree\PhpcrOdmTree::__construct() must be an instance of Symfony\Component\Templating\Helper\CoreAssetsHelper, instance of Symfony\Bundle\FrameworkBundle\Templating\Helper\AssetsHelper given` which was fixed in sonata-project/SonataDoctrinePhpcrAdminBundle#337 but that would bump the requirements to version `1.2.3`. Should I add this?

The `DEPS=dev` build keeps timing out at `composer update --prefer-dist`

**Comment by WouterJ** [comment]
Thanks for putting so much time in this PR! It's great to see Symfony 3 coming closer each push.

 > I tried adding Symfony 3.0.* to the test matrix (https://travis-ci.org/symfony-cmf/RoutingBundle/jobs/105928993) but sonata-project/doctrine-phpcr-admin-bundle 1.1.0 requires symfony/framework-bundle ~2.3 [...] to that fails at the moment.

What you can do is using branch aliases to fool Composer that we're installing 2.8 and then excluding the `web-test` group (using `--exclude-group`) when running PHPunit with Symfony 3. It's a little hacky, but this allows us to see if everything works with Symfony 3 until SonataDoctrinePHPCR properly supports Symfony 3.

 > [...] which was fixed in sonata-project/SonataDoctrinePhpcrAdminBundle#337 but that would bump the requirements to version 1.2.3. Should I add this?

I see no problem bumping that requirement.

 > The DEPS=dev build keeps timing out at composer update --prefer-dist

Composer probably has a hard to resolving the versions. Let's narrow down the search by forcing Symfony 2.8-dev (by also setting `SYMFONY_VERSION="2.8.*"` for this Travis job).

**Comment by WouterJ** [comment]
Btw, I think if we're going to release 2.0 of this bundle, I think we should just remove all BC layers and go with Symfony 2.8. We didn't discuss this much yet, but I suggest to not waste time creating BC layers to support 2.7.

**Comment by hacfi** [comment]
@WouterJ Thanks for the advice! Unfortunately I’m not getting the branch alias working correctly:

With `"symfony/framework-bundle": "v3.0.1 as v2.8.2"` it still says

```
  Problem 1
    - sonata-project/doctrine-phpcr-admin-bundle 1.2.4 requires symfony/framework-bundle ~2.3 -> satisfiable by symfony/framework-bundle[v2.8.2].
    - sonata-project/doctrine-phpcr-admin-bundle 1.2.3 requires symfony/framework-bundle ~2.3 -> satisfiable by symfony/framework-bundle[v2.8.2].
    - sonata-project/doctrine-phpcr-admin-bundle 1.2.4 requires symfony/framework-bundle ~2.3 -> satisfiable by symfony/framework-bundle[v2.8.2].
    - Conclusion: don't install symfony/framework-bundle v2.8.2
    - Installation request for sonata-project/doctrine-phpcr-admin-bundle ^1.1,>=1.2.3 -> satisfiable by sonata-project/doctrine-phpcr-admin-bundle[1.2.3, 1.2.4].
```

Any idea how to get this running?

**Comment by hacfi** [comment]
Prefer lowest was failing because of sonata-project/SonataAdminBundle#2186 so I added `sonata-project/admin-bundle >=2.2.12 `

File: composer.json [hunk 5]
```diff
@@ -23,7 +23,10 @@
         "matthiasnoback/symfony-dependency-injection-test": "~0.6",
         "matthiasnoback/symfony-config-test": "^1.3.1",
         "phpunit/php-code-coverage": "@stable",
-        "sonata-project/doctrine-phpcr-admin-bundle": "^1.1",
+        "sonata-project/admin-bundle": ">=2.2.12",
+        "sonata-project/core-bundle": ">=2.2.3",
+        "sonata-project/block-bundle": ">=2.2.8",
```
Comment (WouterJ) [inline composer.json]: please use `~2.2.8` instead of `>=`. Sonata is very bad in semver

File: composer.json [hunk 6]
```diff
@@ -23,7 +23,10 @@
...
+        "sonata-project/doctrine-phpcr-admin-bundle": "^1.1,>=1.2.3",
```
Comment (WouterJ) [inline composer.json]: this can be changed with `^1.2.3`.

File: composer.json (hunk 5 above)
Comment (hacfi) [inline composer.json]: Yep, I noticed! Should I change `sonata-project/admin-bundle` or `sonata-project/core-bundle` as well? I didn’t want to be to strict as I don’t know the recent development of the packages.

**Comment by dbu** [comment]
+1 for going with 2.7 or 2.8 and removing all BC hacks. also we can remove our own deprecated things but not necessarily in this PR.

**Comment by hacfi** [comment]
So anything else to do here to get this merged? I know the added version constraints are not great but unfortunately they are necessary.

File: composer.json [hunk 7]
```diff
@@ -23,7 +23,10 @@
         "matthiasnoback/symfony-dependency-injection-test": "~0.6",
         "matthiasnoback/symfony-config-test": "^1.3.1",
         "phpunit/php-code-coverage": "@stable",
-        "sonata-project/doctrine-phpcr-admin-bundle": "^1.1",
+        "sonata-project/admin-bundle": ">=2.2.12",
+        "sonata-project/core-bundle": ">=2.2.3",
```
Comment (dbu) [inline composer.json]: please use `^` and not >=, this would allow sonata 3 or 4 or any. (also the line above)

Comment (dbu) [inline composer.json]: the caret operator means this patch version or any minor version higher than this.

**Comment by dbu** [comment]
coming good. we need to create the 1.4 branch before merging this, so that we still can release the 1.4 that supports symfony 2.3

this alone is no reason to jump the branch-alias to 2.0 so we can see when we need to do that.

**Comment by hacfi** [comment]
Updated. I just didn’t want to be too restrictive.

**Comment by hacfi** [comment]
@dbu Anything else to do here? I’d rather remove the Sf27 Compat stuff in a separate PR to not mix things up.

File: .travis.yml [hunk 8]
```diff
@@ -21,11 +21,9 @@ env:
 matrix:
   include:
     - php: 5.6
-      env: DEPS=dev
+      env: DEPS=dev SYMFONY_VERSION=2.8.*
```
Comment (WouterJ) [inline .travis.yml]: if you change this value to `3.0.x-dev as 2.8.99`, we should be able to test against Symfony 3 (I used the same successuflly line in another project)

**Comment by WouterJ** [comment]
@hacfi I think this PR is ready. However, we're focusing on the 1.4 stable release first. After that's done (expect end February at the very last), we start bumping to 2.0 and merging the 2.0 PRs.

**Comment by hacfi** [comment]
That sounds great!

Maybe it would be helpful for others to pull my branch into this repo so they can also use it as `dev-symfony3` - just a suggestion :).

File: .travis.yml (hunk 8 above)
Comment (hacfi) [inline .travis.yml]: Will try this again..thx!

Comment (hacfi) [inline .travis.yml]: Not working: https://travis-ci.org/symfony-cmf/RoutingBundle/jobs/109766483 :(

**Comment by dbu** [comment]
yep, exactly, i want to wait until we have 1.4 released and start the 1.4 branch, then switch master to 2.0 and merge this.

good idea with the branch. that would also allow to do that follow up PR to remove BC hacks - i agree its better to do that separately. i just created a symfony3 branch from master. can you close this PR and open a new one against the symfony3 branch? then we get that merged asap and can continue further steps in separate PRs.

**Comment by hacfi** [comment]
Done: #332

**Pull request closed without merging**